### ISSUE TITLE
Unique DeviceId

### ISSUE THREAD

**Issue by mjaddani** [google/shaka-player#1523]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**: Yes

**What version of Shaka Player are you using?**: 2.4.2

**Can you reproduce the issue with our latest release version?**: Yes

**Can you reproduce the issue with the latest code from `master`?**: Yes

**Are you using the demo app or your own custom app?**: Both

**If custom app, can you reproduce the issue using our demo app?**: not applicable

**What browser and OS are you using?**: Chrome on Windows 10

**What are the manifest and license server URIs?**: We don't need this to reproduce this issue

**What did you do?**
We whant to have à unique deviceId for chrome Browser nad firefox. Widevine advised me to activate these 2 attributes
config.drm.advanced = ({
        'com.widevine.alpha': {
            distinctiveIdentifierRequired: true,
            persistentStateRequired: true,
        },
    });

But if do that, i have this error on shaka player
**Shaka Error DRM.REQUESTED_KEY_SYSTEM_CONFIG_UNAVAILABLE ()**

what is the solution ?

**What did you expect to happen?**

**What actually happened?**

**Comment by theodab** [comment]
Well, trying those configs out on our sample assets...
Narrowing it down, `persistentStateRequired: true` works, but `distinctiveIdentifierRequired: true` does not.

The `REQUESTED_KEY_SYSTEM_CONFIG_UNAVAILABLE` error is thrown if `navigator.requestMediaKeySystemAccess` fails for every DRM system.
Now, `distinctiveIdentifierRequired: true` from our config is translated into `distinctiveIdentifier: "required"` in the final config passed to `navigator.requestMediaKeySystemAccess`. It's possible that `distinctiveIdentifier: "required"` isn't supported anymore, or something. I'll bring this to the attention of the Widevine team.

In the meantime, does your asset play with just `persistentStateRequired: true` set?

**Comment by mjaddani** [comment]
I'have tested "persistentStateRequired: true", but we have some strange issue, If i try to play 10 lives content we have 4 failures, we have a black screen and no error on the console.

Do we need any think on the licence ?

**Comment by theodab** [comment]
Well, I can't say much more without having access to your content. You can send the information (manifest URI, license server URI, etc) to <shaka-player-issues@google.com>, if you are able to distribute them.
If not, you could go to `chrome://media-internals/` while playing in Chrome to see if the video element's log has any errors in it. If Shaka Player isn't showing any errors, there might be a problem somewhere in the media pipeline that we don't get to touch.

**Comment by theodab** [comment]
I tried to reproduce this on Chrome and Firefox on macOS for a while, and was unable to with either.
On a Windows 10 computer, on the other hand, the video was sometimes hanging forever without actually playing. The screen wasn't black, per se, but it might be what you were seeing.

Interestingly, I was getting media internal logs like this (key ID censored for privacy reasons):
```
00:00:02:103 DecryptingDemuxerStream: no key for key ID *KEYID*
00:00:02:103 DecryptingVideoDecoder: no key for key ID *KEYID*
00:00:02:854 DecryptingDemuxerStream: key was added, resuming decrypt
00:00:02:854 DecryptingVideoDecoder: key added, resuming decode
00:00:02:855 DecryptingDemuxerStream: no key for key ID *KEYID*
00:00:02:855 DecryptingVideoDecoder: no key for key ID *KEYID*
```
And then no other updates after that about the decoder or demuxer.

So the key is supposedly added, and then a millisecond later it's saying that there's no key for that key ID again? And then it just is hanging there, waiting for that key. I don't know how the CDM works well enough to say what's happening here; maybe it's getting a bad key or something.

**Comment by TheModMaker** [comment]
The `key was added` log is printed whenever any keys are added, so what is likely happening is we are getting the wrong keys from the license server.  You should check the key status changed event to see what keys were added vs what keys we need.  Also you should check the expiration of the sessions to see if the session is expiring before we can use it.

@mjaddani I think that distinctive identifiers are only supported on Android which is why you are getting an error on desktop platforms.  I don't think there is a way to get unique IDs for desktop since the OS doesn't provide that info.  Unfortunately you'll need to use platform detection to determine if you are on Android and conditionally set that attribute.

**Comment by mjaddani** [comment]
When the playing failed, i'have all the time this.

00:00:00 691 | info | DecryptingDemuxerStream: no key for key ID XXXXXX
00:00:00 691 | info | DecryptingVideoDecoder: no key for key ID XXXXXX
00:00:00 224 | duration | 4294967296
00:00:01 734 | info | DecryptingDemuxerStream: key was added, resuming decrypt
00:00:01 734 | info | DecryptingVideoDecoder: key added, resuming decode
**00:00:01 734 | info | DecryptingDemuxerStream: no key for key ID XXXXXX
00:00:01 735 | info | DecryptingVideoDecoder: no key for key ID XXXXXX**

I'have the same process on proxy server, the same key, and the duration session it's same all the time.

**Comment by mjaddani** [comment]
This case is like #1554 

do you have analyse my content ?

**Comment by joeyparrish** [comment]
It's not clear to me from the issue history what your problem is right now.  Can you please explain it again?  Also, can you please provide some content URLs so that we can reproduce your issue exactly?  Thanks!

**Comment by mjaddani** [comment]
No problem, We use a live content with widevine DRM.

If i use `persistentStateRequired: true` on shaka, most of the time the content works, but some times, shaka don't play the content and we have an error with Firefox and no error with Chrome but the content don't play.

On chrome,  we have this error on` chrome://media-internals/`
```
00:00:00 691 | info | DecryptingDemuxerStream: no key for key ID XXXXXX
00:00:00 691 | info | DecryptingVideoDecoder: no key for key ID XXXXXX
00:00:00 224 | duration | 4294967296
00:00:01 734 | info | DecryptingDemuxerStream: key was added, resuming decrypt
00:00:01 734 | info | DecryptingVideoDecoder: key added, resuming decode
00:00:01 734 | info | DecryptingDemuxerStream: no key for key ID XXXXXX
00:00:01 735 | info | DecryptingVideoDecoder: no key for key ID XXXXXX
```

And for Firefox, we have this
```
{
"severity": 2,
"category": 3,
"code": 3016,
"data": [
3,
null,
"NS_ERROR_DOM_MEDIA_FATAL_ERR (0x806e0005) - class mozilla::ipc::IPCResult __cdecl mozilla::gmp::ChromiumCDMParent::RecvDecodeFailed(const unsigned int &): ChromiumCDMParent::RecvDecodeFailed"
],
"handled": false,
"message": "Shaka Error MEDIA.VIDEO_ERROR (3,,NS_ERROR_DOM_MEDIA_FATAL_ERR (0x806e0005) - class mozilla::ipc::IPCResult _cdecl mozilla::gmp::ChromiumCDMParent::RecvDecodeFailed(const unsigned int &): ChromiumCDMParent::RecvDecodeFailed)",
"stack": "shaka.util.Error@******/lib/util/error.js:88:13\nshaka.Player.prototype.onVideoError@*****/lib/player.js:3114:17\n"
}
```

But if i disable `persistentStateRequired: false`, we don't have any error. We send the same widevine licence, with the same id.

I send you email to reproduct that

**Comment by vaage** [comment]
@mjaddani If I understand correctly, you are looking for a way to track device ids for Chrome and Firefox. Are you looking to track this in the license requests?

When you say "Widevine advised me to activate these 2 attributes", where did you get that instruction from?

**Comment by mjaddani** [comment]
Yes we want to track device Id to do a restriction, we use `persistentStateRequired: true`, and we use `provider_client_token` on widevine licence.

It's Widevine engeneer who has give me this  advice.

**Comment by vaage** [comment]
So Shaka Player is part of Widevine - who in Widevine have you been talking to? We can touch base with them to figure out both ends of this issue because I suspect most of this is license based as the player has no actual control over the CDM.

**Comment by mjaddani** [comment]
On Widevine DRM Proxy documentation, we have this, 

```
Client token to send back in the license response.  If
the license request contains a client token, this value
is ignored. The client token will persist beyond license
sessions.
provider_client_token is only supported in the Chrome
CDM and persistentState must be enabled by the Application.
```

My Widevine contact is Prashant Grover. Also I have met Joey Parrish in London as part of Widevine training

**Comment by vaage** [comment]
I have reached out to Prashant Grover and will figure out what he is suggesting you do. I will assign this to myself so that we don't lose track of it.

**Comment by mjaddani** [comment]
My need is to be able to have a device id

**Comment by vaage** [comment]
After talking this over with Prashant, he said his original instructions would have been:
- Specify Provider Client Token on your proxy.
- Set `distinctiveIdentifierRequired` and `persistentStateRequired` in the player.

As you pointed out,
> provider_client_token is only supported in the Chrome CDM and persistentState must be enabled by the Application.

For distinctive identifier to work, the CDM must be willing to provide one, that is why we query for it. If the configuration cannot be fulfilled (REQUESTED_KEY_SYSTEM_CONFIG_UNAVAILABLE) there is nothing we can do, the EME specs even say,
> Implementations that use Distinctive Identifier(s) or Distinctive Permanent Identifier(s) should support the option to not use them.

And it even recomends against their use ([source](https://www.w3.org/TR/encrypted-media/#limit-or-avoid-use-of-distinctive-identifiers-and-permanent-identifiers)).
 
So @mjaddani, with all that said, unless a platform is willing to give you distinctive identifiers, there is no way to force it.

**Comment by mjaddani** [comment]
Thanks @vaage , but for the Android Device, we have solution. My point is for Chrome and Firefox, the advise it's to use `persistentStateRequired=true `and Widevine Proxy set the param `provider_client_token` with a device Id. This advise work but sometimes I have some mistakes that I went back to above

**Comment by TheModMaker** [comment]
Sorry not following-up sooner.  I believe this is a duplicate of #1554, which has been fixed by Widevine and will appear at least in Chrome 71, maybe sooner.  It was also fixed on the CDM side, so it should appear in Firefox too, although I don't know when.
